### ISSUE TITLE
Add image requirements metadata

### ISSUE THREAD

**Issue by tomponline** [lxc/lxd#9408]
Adding additional image metadata that indicates the requirements the image has, so that LXD can inspect it and check the system and settings for compatibility.

This would be useful for (at least) the following two scenarios:

1. VM images that require secure boot be disabled.
2. Container images that use systemd and require the host to have a supported cgroup configuration.

If LXD could access the requirements of the image it could then detect if there are issues that would prevent the instance from working and flag it to the end user.

Otherwise in both scenarios above it is very difficult for the end user to diagnose the fault.

**Comment by jeff2814** [comment]
Hi, @hQavi and I are students from University of Texas taking a virtualization class. One of our assignments is to contribute to an open source project and we thought lxd was pretty exciting! Would it be ok if we tried to tackle this issue?

**Comment by stgraber** [comment]
Sure! I've assigned it to you for now. I need to have @hQavi comment in here so I can assign them too.

**Comment by stgraber** [comment]
Done!

**Comment by jeff2814** [comment]
Hi, unfortunately we weren't able to tackle to this issue this semester. So it'll be best to let someone else take the issue. 

We also wanted to thank the maintainers for letting us contribute to the lxd! It's been a pleasure working on lxd and we might be back in the future. In the meantime, we wish y'all the best!

**Comment by stgraber** [comment]
Thanks, I've cleared the assignee field.

**Comment by cavcrosby** [comment]
@stgraber could I take a stab at this? I think I have a good diving point into the simplestreams client, probably will have questions as I go along

**Comment by stgraber** [comment]
Sounds good, assigned it to you

**Comment by cavcrosby** [comment]
Had a few hours to chew on things, definitely want to make sure I am on the right track here.

For the simplestreams client, under ```./shared/simplestreams/products.Product``` (at least I believe I think this is where the additional simplestreams keys would go) I was thinking of using a field with the type ```map[string]interface{}``` because each requirement/key could possible require more information aside from what a scalar value could provide. Perhaps this isn't necessary and something more simple could be used instead of an empty interface (e.g. ```string```)? You did mention I can add additional simplestreams keys but if they serve a related purpose, wouldn't it be better to have them under a single map?

To clarify, the purpose of this ticket/issue is to add additional metadata to an image dictating host requirements required before trying to create a container...but would this include adding functionality to LXD for following host requirements mentioned by @tomponline above (e.g. VM images that require secure boot be disabled.)?

**Comment by cavcrosby** [comment]
I believe I have a rough implementation drafted out. Do we want LXD to still be able to init a instance (e.g. create but not start the instance) even if host has a missing requirement based on the new image metadata (e.g. there exist an invalid cgroup configuration on the host)?

**Comment by stgraber** [comment]
> For the simplestreams client, under `./shared/simplestreams/products.Product` (at least I believe I think this is where the additional simplestreams keys would go) I was thinking of using a field with the type `map[string]interface{}` because each requirement/key could possible require more information aside from what a scalar value could provide. Perhaps this isn't necessary and something more simple could be used instead of an empty interface (e.g. `string`)? You did mention I can add additional simplestreams keys but if they serve a related purpose, wouldn't it be better to have them under a single map?

I think adding a `lxd_requirements` as `map[string]string]` should be sufficient here. It should allow for enough flexibility without having to either use a heavy approach with a dedicated struct or the like or one that's hard to parse like using interfaces.

> To clarify, the purpose of this ticket/issue is to add additional metadata to an image dictating host requirements required before trying to create a container...but would this include adding functionality to LXD for following host requirements mentioned by @tomponline above (e.g. VM images that require secure boot be disabled.)?

I'd rather not, having an image server or image be able to cause LXD to change instance configuration, while it sounds great can also very quickly turn into a security issue. I think for now, we're effectively looking at adding this data to simplestreams and our parser, then have LXD pick that up and add some instance startup checks to refuse start with a clear error if something is needed.

**Comment by stgraber** [comment]
As for current implementation, the two things I suspect we'll do are:
 - Label a bunch of container images with `map[string]string{"cgroup": "v1"}` to indicate that they require cgroupv1 and will not work on a system that has a full cgroupv2.
 - Label some VM images with `map[string]string{"secureboot": "false"}` to indicate that the image cannot boot with secureboot enabled.

I'd expect our simplestreams parser to read that `lxd_requirements` map and then when generating the LXD image data, to turn that into properties, basically giving us additional image properties like:
 - requirements.cgroup=v1
 - requirements.secureboot=false

This will then automatically get propagated by LXD from the image to the instance when we copy the image properties.
The instance will therefore end up with those keys in its config:
 - image.restrictions.cgroup=v1
 - image.restrictions.secureboot=false

And we can then easily have our `Start()` function process the requirements and fail startup with a clear error should a requirement not be matched.

**Comment by stgraber** [comment]
We'll need to update `doc/image-handling.md` to specify the list of restrictions that are supported on images and what they mean, but other than that changes required for this should be pretty minimal overall.

**Comment by tomponline** [comment]
@stgraber @monstermunchkin shall we re-open this issue, or create a new issue(s) in order to track adding the new metadata to the appropriate images?

**Comment by stgraber** [comment]
I've got a card tracking adding the metadata to the publishing scripts.
